test/gomod_git_auth_test.go

```diff
@@ -0,0 +1,1067 @@
+		// to privide the private key upon request. We start an agent here to
+		// serve the private key.
+		pubkey, privkey := generateKeyPair(t)
+		agentErrChan := startSSHAgent(t, privkey, sockaddr)
+
+		netHostBuildxEnv.RunTestOptsFirst(ctx, t, []testenv.TestRunnerOpt{
+			// This tells buildkit to forward the SSH Agent socket, giving the
+			// gomod generator worker access to the private key
+			testenv.WithSSHSocket(sshID, sockaddr),
+			testenv.WithHostNetworking,
+		}, func(ctx context.Context, client gwclient.Client) {
+			// This MUST be called at the  start of each test. Because we
+			// persist the go mod cache between runs, the git tag of the
+			// private go module needs to be unique. Within a single run of the
+			// tests, the http and ssh tests cannot have the same tag or we
+			// risk a false positive due to caching.
+			attr := attr.WithNewPrivateGoModuleGitTag()
+
+			testState := TestState{
+				t:       t,
+				ctx:     ctx,
+				_client: client,
+				attr:    *attr,
+			}
+
+			_, repo, gitHost := initStates(&testState)
+			sshGitHost := gitHost.
+				// The SSH host has to know the public key that corresponds to
+				// the client's private key; otherwise it will deny access.
+				With(authorizedKey(pubkey, "/root")).
+				// In order to host a git repo over SSH, it needs to be hosted
+				// as a "bare" git repo. An existing git repo can be made into
+				// a bare repo by hoisting the contents of the `.git` directory
+				// up a level, removing everything else, and calling `git init
+				// --bare`. That bare repo can then be used as a git remote
+				// over SSH.
+				With(bareRepo(repo, attr.RepoAbsDir()))
+
+			const githostUsername = "root"
+			sshErrChan := testState.startSSHServer(sshGitHost)
+
+			spec := testState.generateSpec(depModfileContents(t, attr), dalec.GomodGitAuth{
+				SSH: &dalec.GomodGitAuthSSH{
+					ID:       sshID,
+					Username: githostUsername,
+				},
+			})
+			sr := newSolveRequest(
+				withBuildTarget("debug/gomods"),
+				withSpec(ctx, t, spec),
+				// The extra host in fquestion here is the
+				withExtraHost(testState.attr.PrivateGomoduleHost, testState.attr.GitRemoteAddr),
+			)
+
+			solveResultChan := make(chan *gwclient.Result)
+			solveErrChan := make(chan error)
+
+			solveTCh(ctx, t, testState.client(), sr, solveResultChan, solveErrChan)
+
+			var res *gwclient.Result
+			select {
+			case err := <-agentErrChan:
+				t.Fatalf("ssh agent unexpededly failed: %s", err)
+			case err := <-sshErrChan:
+				t.Fatalf("ssh server unexpectedly failed: %s", err)
+			case err := <-solveErrChan:
+				t.Fatalf("solve failed: %s", err)
+			case r := <-solveResultChan:
+				res = r
+			}
+
+			filename := calculateFilename(ctx, t, attr, res)
+			checkFile(ctx, t, filename, res, []byte("bar\n"))
+		})
+	})
+}
+
+// GitServicesAttributes are the basic pieces of information needed to host two git
+// servers, one via SSH and one via HTTP
+type GitServicesAttributes struct {
```

Comment:
Maybe this along with all the git setup stuff (and things like the `file` type) should go into a separate package so as not to pollute the namespace of this package.